### PR TITLE
[to #67] BR backup&restore for apiv2

### PR DESCRIPTION
**Pull request by haojinming**
Signed-off-by: haojinming <jinming.hao@pingcap.com>

### What problem does this PR solve?

Issue Number: ref #67 

Problem Description:  Backup&Restore on apiv2 fails.

### What is changed and how does it work?
1. support apiv2 backup and restore
2. support backup in apiv1 and restore to apiv2
### Code changes

- Has exported function/method change
- Has exported variable/fields change

### Check List for Tests

This PR has been tested by at least one of the following methods:
- Unit test
- integration test
- Manual test (add detailed scripts or steps below)
- 
### Side effects
- No side effects

### CODE REVIEW

**Review by zz-jason — DISMISSED** [review]
LGTM

File: br/pkg/utils/key.go [hunk 1]
```diff
@@ -89,3 +95,29 @@ func CompareEndKey(a, b []byte) int {
...
+		return []byte{APIV2KeyPrefixEnd}
+	}
+	apiv2Key := []byte{APIV2KeyPrefix}
+	return append(apiv2Key, key...)
+}
+
+// conversion between formated APIVersion key and backupmeta key
+// for example, backup apiv1 -> apiv2, add `r` prefix and `s` for empty end key.
+// apiv2 -> apiv1, remove first byte. This is useful when decode the response keys.
+func ConvertBackupConfigKey(backupKey []byte, isEnd bool, srcAPIVer kvrpcpb.APIVersion, dstAPIVer kvrpcpb.APIVersion) []byte {
```
Comment (pingyu) [inline br/pkg/utils/key.go:111]: How about define a struct as `Range`, and do the codec, and cross api version conversion as a whole ? Then the caller do not need to care about the `isEnd`.

Besides, additional to backup meta, this conversion function can be used in other place. E.g., `cfg.adjustBackupRange` can be treated as a conversion from v1 to source api version.

File: br/pkg/utils/key.go [hunk 2]
```diff
@@ -89,3 +95,29 @@ func CompareEndKey(a, b []byte) int {
...
+// encode
+func FormatAPIV2Key(key []byte, isEnd bool) []byte {
+	if isEnd && len(key) == 0 {
+		return []byte{APIV2KeyPrefixEnd}
+	}
+	apiv2Key := []byte{APIV2KeyPrefix}
+	return append(apiv2Key, key...)
+}
+
+// conversion between formated APIVersion key and backupmeta key
```
Comment (pingyu) [inline br/pkg/utils/key.go:108]:
```suggestion
// ConvertBackupConfigKey converts between formated APIVersion key and backupmeta key
```
It is recommended to comment every function in the style as "`Function-Name` does ...". Please refer to [here](https://go.dev/doc/effective_go#commentary).

Moveover, please add unit tests for new functions.

**Comment by pingyu** [comment]
Please resolve the failure of `br-unit-test` in CI.

File: br/pkg/task/restore_raw.go [hunk 3]
```diff
@@ -61,6 +60,9 @@ func RunRestoreRaw(c context.Context, g glue.Glue, cmdName string, cfg *RestoreR
 	if err != nil {
 		return errors.Trace(err)
 	}
+	// for restore, dst and cur are the same.
+	cfg.DstAPIVersion = backupMeta.ApiVersion.String()
```
[the remark this answers is not in the document]
Comment (haojinming) [reply br/pkg/task/restore_raw.go:64]: this is the restore scenario. Actually, user don't need specify the destination api-version. As we only support restore different backup files into cluster. So the DstAPIVersion  is set after backupMeta is loaded.

File: br/pkg/backup/client.go [hunk 4]
```diff
@@ -129,6 +142,43 @@ func (bc *Client) GetTS(ctx context.Context, duration time.Duration, ts uint64)
...
 }
 
+func (bc *Client) GetCurrentTiKVApiVersion(ctx context.Context) (kvrpcpb.APIVersion, error) {
+	allStores, err := conn.GetAllTiKVStoresWithRetry(ctx, bc.mgr.GetPDClient(), conn.SkipTiFlash)
+	if err != nil || len(allStores) == 0 {
+		return kvrpcpb.APIVersion_V1, errors.Trace(err)
+	} else if len(allStores) == 0 {
+		return kvrpcpb.APIVersion_V1, errors.New("store are empty")
+	}
+	url := fmt.Sprintf("%s://%s/config", "http", allStores[0].StatusAddress)
```
[the remark this answers is not in the document]
Comment (haojinming) [reply br/pkg/backup/client.go:152]: ok

File: br/pkg/utils/key.go (hunk 1 above)
Comment (haojinming) [reply br/pkg/utils/key.go:111]: good idea.

File: br/pkg/task/rawkv_config.go [hunk 5]
```diff
@@ -96,9 +96,11 @@ func (cfg *RawKvConfig) ParseFromFlags(flags *pflag.FlagSet) error {
 		return errors.Annotate(berrors.ErrBackupInvalidRange, "endKey must be greater than startKey")
 	}
 
-	// parse and verify destination API version.
-	if err = cfg.parseDstAPIVersion(flags); err != nil {
-		return err
+	if parseAPIVer {
```
[the remark this answers is not in the document]
Comment (haojinming) [reply br/pkg/task/rawkv_config.go]: resolved.

File: br/pkg/backup/client.go [hunk 6]
```diff
@@ -129,6 +142,43 @@ func (bc *Client) GetTS(ctx context.Context, duration time.Duration, ts uint64)
 	return backupTS, nil
 }
 
+func (bc *Client) GetCurrentTiKVApiVersion(ctx context.Context) (kvrpcpb.APIVersion, error) {
+	allStores, err := conn.GetAllTiKVStoresWithRetry(ctx, bc.mgr.GetPDClient(), conn.SkipTiFlash)
+	if err != nil || len(allStores) == 0 {
```
[the remark this answers is not in the document]
Comment (haojinming) [reply br/pkg/backup/client.go]: resolved.

File: br/pkg/utils/key.go [hunk 7]
```diff
@@ -10,9 +10,15 @@ import (
 	"strings"
 
 	"github.com/pingcap/errors"
+	"github.com/pingcap/kvproto/pkg/kvrpcpb"
 	berrors "github.com/tikv/migration/br/pkg/errors"
 )
 
+const (
+	APIV2KeyPrefix    byte = 'r'
+	APIV2KeyPrefixEnd byte = 's'
```
[the remark this answers is not in the document]
Comment (haojinming) [reply br/pkg/utils/key.go]: resolved.

File: br/pkg/utils/key.go (hunk 1 above)
Comment (haojinming) [reply br/pkg/utils/key.go]: done.

File: br/pkg/utils/key.go (hunk 2 above)
Comment (haojinming) [reply br/pkg/utils/key.go]: done

File: br/pkg/restore/split_test.go [hunk 8]
```diff
@@ -330,7 +330,7 @@ func runTestSplitAndScatterWith(t *testing.T, client *TestClient) {
 	regionSplitter := restore.NewRegionSplitter(client)
 
 	ctx := context.Background()
-	err := regionSplitter.Split(ctx, ranges, rewriteRules, false, func(key [][]byte) {}) // TODO: add test case for "isRawKV=true"
+	err := regionSplitter.Split(ctx, ranges, rewriteRules, true, func(key [][]byte) {}) // TODO: add test case for "isRawKV=true"
```
[the remark this answers is not in the document]
Comment (haojinming) [reply br/pkg/restore/split_test.go]: done

File: br/pkg/restore/split_test.go [hunk 9]
```diff
@@ -274,7 +274,7 @@ func TestScatterFinishInTime(t *testing.T) {
...
```
[the remark this answers is not in the document]
Comment (haojinming) [reply br/pkg/restore/split_test.go:277]: done

File: br/pkg/backup/client.go [hunk 10]
```diff
@@ -129,6 +142,43 @@ func (bc *Client) GetTS(ctx context.Context, duration time.Duration, ts uint64)
 	return backupTS, nil
 }
 
+func (bc *Client) GetCurrentTiKVApiVersion(ctx context.Context) (kvrpcpb.APIVersion, error) {
```
[the remark this answers is not in the document]
Comment (haojinming) [reply br/pkg/backup/client.go:155]: test case is added.

File: br/pkg/task/restore_raw.go (hunk 3 above)
[the remark this answers is not in the document]
Comment (haojinming) [reply br/pkg/task/restore_raw.go:64]: moved parseDstAPIVersion to ParseBackupConfigFromFlags

**Comment by haojinming** [comment]
@pingyu  PTAL~   Unit test and integration tests are added.  apiv1ttl -> apiv2 integration test is commented out as tikv v6.0.0 does not support this scenario. Will enable this case after TiKV BR component support PR is merged.

File: br/pkg/utils/key.go [hunk 11]
```diff
@@ -89,3 +95,48 @@ func CompareEndKey(a, b []byte) int {
 
 	return bytes.Compare(a, b)
 }
+
+type KeyRange struct {
+	Start []byte
+	End   []byte
+}
+
+// encode
```
Comment (pingyu) [inline br/pkg/utils/key.go:104]: This comment is too simple. Please polish it, or just remove it.

File: br/pkg/utils/key.go [hunk 12]
```diff
@@ -89,3 +95,48 @@ func CompareEndKey(a, b []byte) int {
...
+func formatAPIV2Key(key []byte, isEnd bool) []byte {
+	if isEnd && len(key) == 0 {
+		return []byte{APIV2KeyPrefixEnd}
+	}
+	apiv2Key := []byte{APIV2KeyPrefix}
+	return append(apiv2Key, key...)
+}
+
+// FormatAPIV2KeyRange convert user key to APIV2 format.
```
Comment (pingyu) [inline br/pkg/utils/key.go:113]: Suggest to add unit tests for the following two functions.

**Review by pingyu — COMMENTED** [review]
Two tiny suggestions. Rest LGTM~

File: br/pkg/utils/key.go (hunk 12 above)
Comment (haojinming) [reply br/pkg/utils/key.go:112]: test cased are added.

**Review by pingyu — APPROVED** [review]
Great ! LGTM~

**Review by zeminzhou — COMMENTED** [review]
LGTM~

**Review by zeminzhou — APPROVED** [review]

**Pull request merged**